### PR TITLE
Convert TimeArray to QArrays

### PR DESCRIPTION
**Pull request by gautierronan**
On top of https://github.com/dynamiqs/dynamiqs/pull/630.

Conversion of the `time_array.py` file to use QArrays. Nothing super surprising: the main difficulty is that we need to use `isqarraylike` instead of `isinstance(x, QArrayLike)` because we specify list content in the typing.

The following now runs without erroring:
```python
import dynamiqs as dq
import jax.numpy as jnp

H = dq.number(8)
H += dq.modulated(jnp.sin, dq.destroy(8) + dq.create(8))
jump_ops = [dq.destroy(8)]
tlist = jnp.linspace(0.0, 1.0, 10)
psi0 = dq.basis(8, 0)
result = dq.mesolve(H, jump_ops, psi0, tlist)
```

### CODE REVIEW

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -362,11 +364,11 @@ def broadcast_to(self, *new_shape: int) -> TimeArray:
     def conj(self) -> TimeArray:
         return PWCTimeArray(self.times, self.values.conj(), self.array.conj())
 
-    def __call__(self, t: ScalarLike) -> Array:
-        def _zero(_: float) -> Array:
+    def __call__(self, t: ScalarLike) -> QArray:
+        def _zero(_: float) -> QArray:
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:368]: This method should return a `Qarray`.

File: dynamiqs/time_array.py [hunk 2]
```diff
@@ -529,40 +531,40 @@ def broadcast_to(self, *new_shape: int) -> TimeArray:
     def conj(self) -> TimeArray:
         return SummedTimeArray([tarray.conj() for tarray in self.timearrays])
 
-    def __call__(self, t: ScalarLike) -> Array:
+    def __call__(self, t: ScalarLike) -> QArray:
         return jax.tree_util.tree_reduce(
-            jnp.add, [tarray(t) for tarray in self.timearrays]
+            lambda x, y: x + y, [tarray(t) for tarray in self.timearrays]
         )
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:537]: Does this return a `QArray`?

File: dynamiqs/qarrays/dense_qarray.py [hunk 3]
```diff
@@ -116,11 +114,11 @@ def __repr__(self) -> str:
     def __mul__(self, y: QArrayLike) -> QArray:
         super().__mul__(y)
 
-        if isinstance(y, get_args(ScalarLike)):
+        if isinstance(y, ScalarLike):
```
Comment (pierreguilmin) [inline dynamiqs/qarrays/dense_qarray.py:117]: Why removing `get_args`?
```python
>>> from jaxtyping import ScalarLike
>>> isinstance(1, ScalarLike)
TypeError: Subscripted generics cannot be used with class and instance checks
```

File: dynamiqs/qarrays/qarray.py [hunk 4]
```diff
@@ -327,7 +328,7 @@ def __neg__(self) -> QArray:
     @abstractmethod
     def __mul__(self, y: QArrayLike) -> QArray:
         """Element-wise multiplication with a scalar or an array."""
-        if not isinstance(y, get_args(ScalarLike)):
+        if not isinstance(y, ScalarLike) and not isinstance(y, DynamicJaxprTracer):
```
Comment (pierreguilmin) [inline dynamiqs/qarrays/qarray.py:331]: Could you explain this?

**Review by pierreguilmin — DISMISSED** [review]
Minor changes. Could you also review all the arithmetic methods by keeping in mind that `QArrayLike` can now be e.g. a list of list? (so for example `-y` doesn't work).

**Comment by pierreguilmin** [comment]
Also we should rename this to `TimeQArray`, in another PR to ease the review.

**Comment by pierreguilmin** [comment]
> Also we should rename this to `TimeQArray`, in another PR to ease the review.

And also rename the attributes like `array` to `qarray`.

**Review by abocquet — APPROVED** [review]

**Pull request merged**